File: Examples/Python/BasicUsage/basic_usage.py
import json
import subprocess
import os
from datetime import datetime

class FlexonHelper:
    @staticmethod
    def run_command(args):
        try:
            result = subprocess.run(['flexon-cli'] + args.split(), 
                                  capture_output=True, text=True, check=True)
            return result.stdout
        except subprocess.CalledProcessError as e:
            print(f"Error running Flexon command: {e.stderr}")
            raise

def main():
    print("Flexon Basic Usage Example (Python)")
    print("===================================\n")

    # Create sample data
    data = {
        "user": {
            "name": "Alice Smith",
            "age": 28,
            "email": "alice@example.com",
            "interests": ["programming", "AI", "data science"],
            "metadata": {
                "created_at": datetime.utcnow().isoformat(),
                "last_login": datetime.utcnow().isoformat(),
                "preferences": {
                    "theme": "dark",
                    "notifications": True
                }
            }
        }
    }

    # Save to JSON
    json_file = "user.json"
    flexon_file = "user.flexon"
    encrypted_file = "user_encrypted.flexon"
    decrypted_file = "user_decrypted.json"

    print("1. Creating JSON file...")
    with open(json_file, 'w') as f:
        json.dump(data, f, indent=2)

    # Basic serialization
    print("2. Converting to Flexon format...")
    FlexonHelper.run_command(f"serialize -i {json_file} -o {flexon_file}")

    # Encryption
    print("3. Creating encrypted Flexon file...")
    FlexonHelper.run_command(f"serialize -i {json_file} -o {encrypted_file} -e mysecretkey")

    # Decryption
    print("4. Decrypting Flexon file...")
    FlexonHelper.run_command(f"deserialize -i {encrypted_file} -o {decrypted_file} -e mysecretkey")

    # Verify results
    print("\nResults:")
    print(f"Original JSON size: {os.path.getsize(json_file)} bytes")
    print(f"Flexon size: {os.path.getsize(flexon_file)} bytes")
    print(f"Encrypted size: {os.path.getsize(encrypted_file)} bytes")

    # Compare contents
    with open(json_file, 'r') as f:
        original = json.load(f)
    with open(decrypted_file, 'r') as f:
        decrypted = json.load(f)

    print("\nData verification:")
    print(f"Name matches: {original['user']['name'] == decrypted['user']['name']}")
    print(f"Age matches: {original['user']['age'] == decrypted['user']['age']}")
    print(f"Email matches: {original['user']['email'] == decrypted['user']['email']}")
    print(f"Interests match: {original['user']['interests'] == decrypted['user']['interests']}")

if __name__ == "__main__":
    main()

File: Examples/JavaScript/BasicUsage/basic_usage.js
const fs = require('fs').promises;
const { execSync } = require('child_process');
const path = require('path');

class FlexonHelper {
    static runCommand(args) {
        try {
            return execSync(`flexon-cli ${args}`, { encoding: 'utf8' });
        } catch (error) {
            console.error(`Error running Flexon command: ${error.stderr}`);
            throw error;
        }
    }
}

async function main() {
    console.log("Flexon Basic Usage Example (JavaScript)");
    console.log("======================================\n");

    // Create sample data
    const data = {
        user: {
            name: "Alice Smith",
            age: 28,
            email: "alice@example.com",
            interests: ["programming", "AI", "data science"],
            metadata: {
                created_at: new Date().toISOString(),
                last_login: new Date().toISOString(),
                preferences: {
                    theme: "dark",
                    notifications: true
                }
            }
        }
    };

    // Save to JSON
    const jsonFile = "user.json";
    const flexonFile = "user.flexon";
    const encryptedFile = "user_encrypted.flexon";
    const decryptedFile = "user_decrypted.json";

    console.log("1. Creating JSON file...");
    await fs.writeFile(jsonFile, JSON.stringify(data, null, 2));

    // Basic serialization
    console.log("2. Converting to Flexon format...");
    FlexonHelper.runCommand(`serialize -i ${jsonFile} -o ${flexonFile}`);

    // Encryption
    console.log("3. Creating encrypted Flexon file...");
    FlexonHelper.runCommand(`serialize -i ${jsonFile} -o ${encryptedFile} -e mysecretkey`);

    // Decryption
    console.log("4. Decrypting Flexon file...");
    FlexonHelper.runCommand(`deserialize -i ${encryptedFile} -o ${decryptedFile} -e mysecretkey`);

    // Verify results
    console.log("\nResults:");
    const stats = await Promise.all([
        fs.stat(jsonFile),
        fs.stat(flexonFile),
        fs.stat(encryptedFile)
    ]);

    console.log(`Original JSON size: ${stats[0].size} bytes`);
    console.log(`Flexon size: ${stats[1].size} bytes`);
    console.log(`Encrypted size: ${stats[2].size} bytes`);

    // Compare contents
    const [original, decrypted] = await Promise.all([
        fs.readFile(jsonFile, 'utf8'),
        fs.readFile(decryptedFile, 'utf8')
    ]);

    const originalObj = JSON.parse(original);
    const decryptedObj = JSON.parse(decrypted);

    console.log("\nData verification:");
    console.log(`Name matches: ${originalObj.user.name === decryptedObj.user.name}`);
    console.log(`Age matches: ${originalObj.user.age === decryptedObj.user.age}`);
    console.log(`Email matches: ${originalObj.user.email === decryptedObj.user.email}`);
    console.log(`Interests match: ${JSON.stringify(originalObj.user.interests) === JSON.stringify(decryptedObj.user.interests)}`);

    // Test different encryption algorithms
    console.log("\n5. Testing different encryption algorithms...");
    
    // ChaCha20
    console.log("\nTesting ChaCha20...");
    FlexonHelper.runCommand(`serialize -i ${jsonFile} -o user_chacha20.flexon -e mysecretkey ChaCha20`);
    
    // TripleDES
    console.log("Testing TripleDES...");
    FlexonHelper.runCommand(`serialize -i ${jsonFile} -o user_tripledes.flexon -e mysecretkey TripleDES`);

    // Compare file sizes
    const encryptedStats = await Promise.all([
        fs.stat('user_chacha20.flexon'),
        fs.stat('user_tripledes.flexon')
    ]);

    console.log("\nFile size comparison:");
    console.log(`Original JSON: ${stats[0].size} bytes`);
    console.log(`AES-256: ${stats[2].size} bytes`);
    console.log(`ChaCha20: ${encryptedStats[0].size} bytes`);
    console.log(`TripleDES: ${encryptedStats[1].size} bytes`);
}

main().catch(console.error);
